### PR TITLE
[SPARK-33573][CORE][FOLLOW-UP] Enhance ignoredBlockBytes in pushMergeMetrics to cover more scenarios 

### PR DESCRIPTION
**Pull request by rmcyang**
### What changes were proposed in this pull request?

Currently, the `ignoredBlockBytes` of server side metrics for push-based shuffle does not fully capture the block bytes that were received by ESS but didn't get merged. This PR tries to enhance the logic of  incrementing `ignoredBlockBytes`, to capture those bytes being ignored by ESS. Specifically, the bytes being considered as ignored should be 1. received after the shuffle file is finalized; 2. when a request is for a duplicate block; 3. the bytes ESS received but failed to write.

### Why are the changes needed?

This would enhance the `ignoredBlockBytes` of server side metrics for push-based shuffle to be more accurate.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Modified existing UTs to capture `ignoredBlockBytes`.

### CODE REVIEW

**Comment by rmcyang** [comment]
cc @mridulm @otterc Please take a look, thanks.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (otterc) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1362]: nit: rename to `updatedIgnoreBytesWithDeferredBufs` or `deferredBufsIgnored` or something similar.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (otterc) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1411]: Even when the server aborts, it ignores the bytes.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (otterc) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1365]: I think we should use `buf.remaining` here

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (rmcyang) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java]: Thanks for the suggestions, Chandni. Renamed it to `updateIgnoredBytesWithDeferredBufs` based on the former, as it sounds more smooth to me. Hopefully it looks better. Please let me know otherwise.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (rmcyang) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1413]: Make sense, have updated the PR to call `updateIgnoredBytesWithDeferredBufs()` in `abortIfNecessary()`.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (rmcyang) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java]: Sounds good,  have updated the PR.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (otterc) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1413]: Here it should be `buf.remaining`

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (otterc) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1395]: When it comes here because of `isTooLateBlockPush` then the whole block is going to be ignored. That means the deferredBufs which are getting freed will be ignored as well so we need to add those bytes to the ignoreBlockBytes  as well, correct?

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (otterc) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1371]:
```suggestion
    private void updateIgnoredBytes(long numBytes) {
      if (numBytes > 0) {
        mergeManager.pushMergeMetrics.ignoredBlockBytes.mark(numBytes);
      }
    }

```

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (otterc) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1319]:
```suggestion
        updateIgnoredBytes(freeDeferredBufs());
```

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (otterc) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1398]:
```suggestion
          long deferredBytes = freeDeferredBufs();
          if (isTooLateBlockPush) {
            mergeManager.pushMergeMetrics.lateBlockPushes.mark();
            updateIgnoredBytes(deferredBytes + buf.remaining());
```

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (otterc) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1410]:
```suggestion
            updateIgnoredBytes(freeDeferredBufs() + buf.remaining());
```

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (otterc) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1480]:
```suggestion
          updateIgnoredBytes(freeDeferredBufs());
```

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (otterc) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1500]:
```suggestion
          updateIgnoredBytes(freeDeferredBufs());
```

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (otterc) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1541]:
```suggestion
          updateIgnoredBytes(freeDeferredBufs());
```

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (otterc) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1315]: Wasn't able to comment on the `freeDeferredBufs()`  lines since it is not changed. We can change it to below to apply my other suggestions.
```
    /**
     * @return total number of deferred bytes
     */
    private long freeDeferredBufs() {
      if (deferredBufs == null || deferredBufs.isEmpty()) {
        deferredBufs = null;
        return 0;
      }
      long totalSize = 0;
      for (ByteBuffer deferredBuf : deferredBufs) {
        totalSize += deferredBuf.limit();
        mergeManager.pushMergeMetrics.deferredBlocks.mark(-1);
      }
      mergeManager.pushMergeMetrics.deferredBlockBytes.dec(totalSize);
      deferredBufs = null;
      return  totalSize;
    }
```

Comment (rmcyang) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1321]: This helps reduce iteration on `deferredBufs` which looks better. Have updated the PR, thanks for the suggestions.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (rmcyang) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1415]: Fixed.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (rmcyang) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java]: I think you are right. Have accommodated the suggestions below hence in this case it would be
`updateIgnoredBytes(deferredBytes + buf.remaining())`

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (mridulm) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1368]: `Long` -> `long`

**Review by mridulm — COMMENTED** [review]
I would propose the following change.

Sum up the total number of bytes received in `onData` - does not matter if it was written, deferred, etc.
In onComplete/onFailure, add it to ignored bytes if we are deciding not to use data from this push.
Will keep the change much more constrained, and avoid issues where a push failed/was aborted/etc after it has partially written data out.

Thoughts ?

**Comment by otterc** [comment]
> Sum up the total number of bytes received in `onData` - does not matter if it was written, deferred, etc.
> In onComplete/onFailure, add it to ignored bytes if we are deciding not to use data from this push.
> Will keep the change much more constrained, and avoid issues where a push failed/was aborted/etc after it has partially written data out.

So will we account for bytes that the server failed to write as well in `ignoredBytes`? I was thinking about this while making the suggestions and decided that the bytes that the server fails to write (throws an exception) are **not** `ignored`. 
In the current PR, even when it aborts,  only deferred bytes are included in `ignoredBytes`. The bytes during which the server encounters IOException while writing are not included in `ignoredBytes`.

I am actually okay with your suggestion but I just want to be clear on what we are capturing as ignored. If we are capturing the failed bytes as ignored, then maybe we should either rename it or at least add the comment to make it clear.

**Comment by mridulm** [comment]
Good callout - I do not want to include failed writes, just those which were actually ignored.
We can try/finally the onData, and ignore update in case of IOException

**Comment by otterc** [comment]
If we don't want to include failed writes, then I think your suggestion requires much more refactoring of the current code. I thought it was slightly riskier to make those changes just because of the metric update. I may be wrong. @rmcyang can look into it and see if it is not too much.

**Comment by mridulm** [comment]
Deferred block writes makes it tricker, yes.

**Comment by rmcyang** [comment]
Thanks @otterc and @mridulm !
Have updated the PR trying to cover below proposal:
- For `onComplete` and `onFailure`, add the data not being used from the push to ignored bytes
- For `onData`, add the received data not being used to ignored bytes (include the bytes during which the server encounters IOException)
-  updated the documents for `ignoredBytes` to be clearer on what it tries to collect
Please let me know if you if any thoughts.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (otterc) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1316]: revert this change

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (otterc) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1324]: Revert
```suggestion
        freeDeferredBufs();
```

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (otterc) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1373]:
```suggestion
    private long totalBytesReceived = 0;

    /**
     * Update ignoredBlockBytes in pushMergeMetrics.
     */
    private void receivedBytesIgnored() {
      if (totalBytesReceived > 0) {
        mergeManager.pushMergeMetrics.ignoredBlockBytes.mark(totalBytesReceived);
        totalBytesReceived = 0;
      }
    }

```

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (otterc) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1375]:
```suggestion
    public void onData(String streamId, ByteBuffer buf) throws IOException {
         totalBytesReceived += buf.remaining();

```

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (otterc) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1397]: Revert
```suggestion
          freeDeferredBufs();
```

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (otterc) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1410]: Revert
```suggestion
            freeDeferredBufs();
```

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (otterc) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1431]: Revert this change
```suggestion

```

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (otterc) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1483]: Revert
```suggestion
          freeDeferredBufs();
```

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (otterc) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1490]: Revert 
```suggestion
          freeDeferredBufs();
```

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (otterc) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1502]:
```suggestion
            freeDeferredBufs();
            // Since we don't throw an exception and just return, but the received bytes are ignored we need to update the ignored bytes metric.
             receivedBytesIgnored();
```

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (otterc) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1542]: Revert 
```suggestion
          freeDeferredBufs();
```

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (otterc) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1576]:
```suggestion
      receivedBytesIgnored();
```

**Comment by rmcyang** [comment]
Have updated the PR to reflect the suggestions above and refactored a bit. Thanks @otterc, please let me know if anything is missing.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (mridulm) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1496]: `updateIgnoredBlockBytes` for `isStale` and `isTooLate` conditions above ?

**Comment by mridulm** [comment]
+CC @xinrong-meng, I plan to include this in 3.4 as well.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (otterc) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1496]: We throw the exception so `onFailure` will be called and they will be updated there.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (otterc) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1578]: Nit: Can we move this block to line 1559.
Also reword: `The block was received by ESS but didn't get merged so it is considered as "ignored"`.

**Review by otterc — COMMENTED** [review]
Just have some nits. Please make sure to validate the ignored metric in the UTs in all the scenarios.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (mridulm) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1496]: It need not necessarily be detected during onData - it can go stale subsequently as well.
(Else we wont need these two conditions here at all :-) )

Comment (rmcyang) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1498]: The UTs `testBlockPushWithOlderShuffleMergeId` and `testFinalizeWithOlderShuffleMergeId ` have some stale blocks. Before applying updateIgnoredBlockBytes for isStale and isTooLate conditions above, they are not taken account into the ignored bytes. They are able to be captured after. So I think we should update ignored block bytes there. Updated the PR, thanks.

Comment (otterc) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1498]: @mridulm I didn't get your point. I thought you pointed that in `onComplete` when we detect `tooLate` or `tooStale`, we should update the metric. However, when we detect `tooLate` or `tooStale` we throw an exception, so that will trigger `onFailure`.

@rmcyang Can you check if those tests trigger `onFailure` after `onComplete` throws an exception? If they don't then that's why the metric didn't get updated.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (otterc) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1489]: This is not needed. There is an exception thrown in the next line which triggers `onFailure`

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (otterc) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1481]: This is not needed. There is an exception thrown in the next line which triggers onFailure

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (mridulm) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1498]: >  However, when we detect tooLate or tooStale we throw an exception, so that will trigger onFailure.

I see what you mean, I read it as this wont happen and should have triggered onFailures from onData directly.

Agree - since `TransportRequestHandler.processStreamRequest` is explicitly doing the delegation to `onFailures` when `onComplete` fails, that should be fine.

Comment (rmcyang) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:1498]: @otterc I checked those UTs. We didn't call onFailure explicitly for those steams, I have explicitly call onFailure now and the ignored bytes metric is able to be updated.

**Review by mridulm — COMMENTED** [review]
Just one comment, rest looks good to me.

File: docs/monitoring.md
Comment (rmcyang) [reply docs/monitoring.md]: Thanks for the suggestion, also trimmed a little bit for other metrics above to be consistent :-)

**Review by otterc — APPROVED** [review]
LGTM

**Review by mridulm — APPROVED** [review]

**Comment by mridulm** [comment]
Will wait for CI to pass before merging.

**Comment by mridulm** [comment]
Merged to master and branch-3.4
Thanks for working on this @rmcyang !
Thanks for the review @otterc :-)

+CC @xinrong-meng as this was merged to 3.4 branch

**Comment by rmcyang** [comment]
Thank you for the reviews, @otterc and @mridulm !

**Pull request closed without merging**
